### PR TITLE
Registry lookup for overridden DefaultBean types does not work

### PR DESCRIPTION
**Pull request by JiriOndrusek**
fixes https://github.com/apache/camel-quarkus/issues/3797

This is work in progress/draft to verify that this change does not affect any existing junit tests.

### CODE REVIEW

**Comment by JiriOndrusek** [comment]
Hi @jamesnetherton , can you pls look at this, whether it makes sense to you (it is nut urgent, I'm just tagging you here)

**Comment by jamesnetherton** [comment]
I think it's probably ok. I've been trying to think of scenarios where the way we handle ambiguous beans could cause issues or lead to confusion, but I can't think of any.

Feel free to take the PR out of draft. I'd prefer to have the thoughts of others before merging though.

**Comment by JiriOndrusek** [comment]
Hi @ppalaga, what do you think about this change?

File: integration-test-groups/foundation/bean/src/main/java/org/apache/camel/quarkus/component/bean/BeanResource.java [hunk 1]
```diff
@@ -164,4 +169,18 @@ String awaitFirst(String key) {
...
+    public int withDefaultBeanCount() {
+        return camelContext.getRegistry().findByType(Producers.WithDefaultBeanInstance.class).size();
+    }
+
+    @Path("/withAlternativeBeanCount")
+    @GET
+    @Produces(MediaType.TEXT_PLAIN)
+    public int withAlternativeBeanCount() {
+        return camelContext.getRegistry().findByType(Producers.WithAlternateBeanInstance.class).size();
+    }
```
Comment (ppalaga) [inline integration-test-groups/foundation/bean/src/main/java/org/apache/camel/quarkus/component/bean/BeanResource.java:184]: I see that this primarily covers the cause of the original bug report. But wouldn't it be better to make sure that the registry returns the right single bean returned by the right producer method?

**Review by ppalaga — COMMENTED** [review]
The ambiguity resolution fix looks good. I am just suggesting to be more paranoid in the test.

File: integration-test-groups/foundation/bean/src/main/java/org/apache/camel/quarkus/component/bean/BeanResource.java (hunk 1 above)
Comment (JiriOndrusek) [reply integration-test-groups/foundation/bean/src/main/java/org/apache/camel/quarkus/component/bean/BeanResource.java:184]: That makes sense. thanks for the point, I'll change it/

Comment (JiriOndrusek) [reply integration-test-groups/foundation/bean/src/main/java/org/apache/camel/quarkus/component/bean/BeanResource.java]: - I fixed tests to control also bean instances 
- I added test to cover scenario with 2 beans (without default and alternatives)
- I added test to cover scenario with more beans (from all tests covered by this PR) - 1 of them is default and 1 of them is alternative - to cover back-compatibility on ambiguous failure

**Review by ppalaga — APPROVED** [review]
Perfect, thanks!

**Review by aldettinger — APPROVED** [review]
The failing CI tests seems unrelated. Nice work @JiriOndrusek.

**Review by jamesnetherton — APPROVED** [review]

**Pull request merged**